### PR TITLE
Sign patch id instead of commit hash

### PR DESCRIPTION
**Pull request by tylerlevine**
* also adds github webhook triggering via tag update hack

Signed-off-by: Tyler Levine <tyler@bitgo.com>

### CODE REVIEW

**Comment by tylerlevine** [comment]
This breaks tests because signing the genesis commit is poorly supported. Will fix later.

**Comment by lrvick** [comment]
Signing the genesis commit is an interesting edge case in your scheme, but hey, tests did their job :D

**Comment by KellerFuchs** [comment]
Kinda-dumb question: why is this PR targetting `mac-compatibility` rather than `master`?

File: .gitsigners [hunk 1]
```diff
@@ -8,3 +8,5 @@ A251FDF79171F98674EB2176FCC2D6E33BA86209
 6B61ECD76088748C70590D55E90A401336C8AAA9
 5FEB0160BF6599220E24CE691707EC66F4E06216
 8D83E1C130DF00669E94A5590C7DEDFCED8ECF33
+C9900735BDF49B3F83F6D8EBCC8F1215664DDC11
+
```
Comment (KellerFuchs) [inline .gitsigners]: There are now 2 blank lines at the end of file.
That should also probably be a separate commit from the whitespace fixup.

File: bin/git-signatures [hunk 2]
```diff
@@ -10,6 +10,8 @@ GETOPT=$(command -v gnu-getopt || command -v getopt)
 DATE=$(command -v gdate || command -v date)
 SHUF=$(command -v gshuf || command -v shuf)
 
+SIGNATURE_TAG="latest-signature"
```
Comment (KellerFuchs) [inline bin/git-signatures]: I'm not super-keen on using a tag for this, because there are situations where you might want to have several branches to which you merche things.

For instance, it's very common for software which has several major versions maintained to cut branches like `stable-2.0`, `stable-2.1`, ... for bugfix releases within those series, and merge new features to `master` (which is then split into a new `stable-*` branch at release-time)

Comment (KellerFuchs) [reply bin/git-signatures]: PS: Can't we keep the previous `merge-base` behaviour there?

File: bin/git-signatures [hunk 3]
```diff
@@ -215,6 +217,22 @@ path_check() {
...
+		echo "4b825dc642cb6eb9a060e54bf8d69288fbee4904";
+	else
+		echo "HEAD^"
+	fi
+}
+
+sig_subject() {
+	git diff-tree -p "$1"..HEAD \
+	| git patch-id --stable \
+	| cut -d\  -f1
```
Comment (KellerFuchs) [inline bin/git-signatures]: `cut -d ' ' -f 1` might be clearer

File: bin/git-signatures [hunk 4]
```diff
@@ -215,6 +217,22 @@ path_check() {
...
+	if [ "$?" -eq 128 ]; then
+		echo "4b825dc642cb6eb9a060e54bf8d69288fbee4904";
+	else
+		echo "HEAD^"
+	fi
+}
+
+sig_subject() {
+	git diff-tree -p "$1"..HEAD \
+	| git patch-id --stable \
```
Comment (KellerFuchs) [inline bin/git-signatures]: It's not clear that `patch-id` is collision-resistant.  From the documentation:
> A "patch ID" is nothing but a sum of SHA-1 of the file diffs associated with a patch [...]
> two patches that have the same "patch ID" are almost guaranteed to be the same thing.

I will probably have to check the source to see what they are doing.

**Comment by KellerFuchs** [comment]
@tylerlevine Thanks a lot for the PR, it's a really cool idea.
I left you a bunch of actionable feedback; sorry it took so long before I could make the time to review.  :3

File: bin/git-signatures (hunk 3 above)
Comment (RyanSquared) [reply bin/git-signatures]: I wouldn't mind `cut -d' ' -f1` either.

File: bin/git-signatures (hunk 2 above)
Comment (lrvick) [reply bin/git-signatures]: This is a hack because github won't notice new signatures otherwise. Got a better one?

File: bin/git-signatures (hunk 3 above)
Comment (KellerFuchs) [reply bin/git-signatures]: Works too, yes.

File: bin/git-signatures (hunk 2 above)
Comment (KellerFuchs) [reply bin/git-signatures]: Wait, that's independent of the patch-id signing?  Can we move that to a separate issue/PR, then?

**Comment by lrvick** [comment]
I think we need to allow the future ability to disallow out of order merges as that can have security consequences. To do this we would need the current master head commit hash to be both inside and outside the signature for reference so future git-signatures can flag if a diff was applied earlier or later than intended.

Other than that and kellers suggestions I am +1

**Comment by tylerlevine** [comment]
> Kinda-dumb question: why is this PR targetting `mac-compatibility` rather than `master`?

Only because this branch depended on changes in the mac-compatibility branch when it was created, and then that branch was merged into master. I guess github isn't smart enough to update the PR target if that ref is merged somewhere. I'm expecting that this PR will land in master now.

File: .gitsigners (hunk 1 above)
Comment (tylerlevine) [reply .gitsigners]: Yeah you're right, I'll rip this out

File: bin/git-signatures (hunk 2 above)
Comment (tylerlevine) [reply bin/git-signatures]: > I'm not super-keen on using a tag for this, because there are situations where you might want to have several branches to which you merche things.
> 
> For instance, it's very common for software which has several major versions maintained to cut branches like stable-2.0, stable-2.1, ... for bugfix releases within those series, and merge new features to master (which is then split into a new stable-* branch at release-time)

As @lrvick alluded to, I don't really care about what this tag points to, I only care that it gets moved somewhere else when a signature is updated. I'm just using the tag update as a mechanism for getting github to fire a webhook.

> Wait, that's independent of the patch-id signing? Can we move that to a separate issue/PR, then?

Yeah, it doesn't depend on the patch signing. I'll rip this out too

**Comment by tylerlevine** [comment]
> I think we need to allow the future ability to disallow out of order merges as that can have security consequences. To do this we would need the current master head commit hash to be both inside and outside the signature for reference so future git-signatures can flag if a diff was applied earlier or later than intended.
> 
> Other than that and kellers suggestions I am +1

I've already started working on that. Essentially this changes the meaning of the review signature from acceptance of a certain patch, to acceptance of a certain patch relative to another commit. If the author tries to land an accepted patch on a commit other than the one approved by the reviewer, the review signature would be invalid. I think that's pretty cool.

**Comment by KellerFuchs** [comment]
@lrvick Yes, we should enforce that the merges stack up properly; that possibly involves putting the base commit hash into the signature, in addition to the patch id?

**Comment by lrvick** [comment]
@KellerFuchs Yeah pretty much just need to sign the patch-id/master-head combo but leave them in the clear so then a future flag on "git signatures verify" will have the information it needs to be opt in/out of that level of strictness. It does mean you can't really have more than one ready-to-merge PR at a time so people may choose to disable this level of strictness on less sensitive repos. 

Tradeoffs.

File: bin/git-signatures (hunk 4 above)
Comment (KellerFuchs) [reply bin/git-signatures]: Yes; my fear was that if “a sum of SHA-1” is really an addition (on 160b numbers?) rather than hashing those together, it might be terrifyingly easy to collide.

Comment (KellerFuchs) [reply bin/git-signatures]: I followed the calls (reading the source) until I hit [`diff_get_patch_id`](https://github.com/git/git/blob/7e8bfb0412581daf8f3c89909f1d37844e8610dd/diff.c#L5633-L5723) which seems to be doing the right thing.

The collision resistance is still that of SHA1, but that ought to be Good Enough © for now.

**Comment by KellerFuchs** [comment]
Oh, of course it doesn't show after rebase, even though the diff is the same:
```
$ ./bin/git-signatures show
 Public Key ID    | Status     | Trust     | Date                         | Signer Name                                        
=======================================================================================================================
error: no note found for object 0d0f7ac917b3bfae9b1f5ee308866494d9206d19.
```

**Comment by tylerlevine** [comment]
> Oh, of course it doesn't show after rebase, even though the diff is the same:
> 
> ```
> $ ./bin/git-signatures show
>  Public Key ID    | Status     | Trust     | Date                         | Signer Name                                        
> =======================================================================================================================
> error: no note found for object 0d0f7ac917b3bfae9b1f5ee308866494d9206d19.
> ```

Try rerunning `git signatures init` so that `notes.rewriteRef` gets set properly. Also I don't have the pubkey you used, here is the one I have for you (which is also present in .gitsigners):

```
pub   rsa4096/0527B4EFFB4A3AEB 2017-10-31 [C] [expires: 2020-10-30]
      AE2D535ABD2E5B42CE1E97110527B4EFFB4A3AEB
uid                 [ultimate] Keller Fuchs
uid                 [ultimate] kellerfuchs@hashbang.sh
sub   rsa4096/7AE119B545B40B2A 2017-10-31 [S] [expires: 2018-10-31]
```

Pubkey used to sign was instead FA1FC72F261DC619

**Comment by KellerFuchs** [comment]
```
$ ./bin/git-signatures init
$ ./bin/git-signatures show
 Public Key ID    | Status     | Trust     | Date                         | Signer Name                                        
=======================================================================================================================
error: no note found for object 0d0f7ac917b3bfae9b1f5ee308866494d9206d19.
```

Re: signing key, indeed somehow only my rsa signing key got added...

**Comment by lrvick** [comment]
So I have been thinking a lot on this, and realized there is already yet one more thing we will want to potentially attest to in a future: The compiled binary hash this given patch-id produces.

So far the things that make sense to sign are:
* patch-id (always)
* target head (optional)
* compiled binary hash (optional)

There may yet be others we have not considered yet. 

For this reason since we are making a breaking change to the signing format with this commit, let's tweak it to be future proof by using key/value pairs. 

JSON sounds the easiest for this, but that will require we depend on jq (parsing json in bash in hell). Jq is becoming very common though so maybe this is acceptable until someone is willing to rewrite this whole thing as a Rust binary or somthing.

Anyway I was thinking a signature could then just be a single line of json containing a body and a signature for that body

{body:{ patchid:"somepatchid",bin-hash:"somebinhash",target-head:"targetheadhere"},signature:"owGbwMvMwMHY577wjenCUFnG00FJDNFzp3pWFySWJGdkplgpFefnpkI5SjpJmXm6GYnFGRBhIA/E
UdIpSSxKTy3RzUhNBGqAcEDsjNSiVKVark4mYxYGRg4GWTFFlvRQ+71LsrdL7dXj5oZZysoEspGB
i1MAJsIzkf8P37uGc1wx5SJH0kK6uGU+Lbu2WVtq9brA62/f5arGnBG4ct37pEa9++LuTZMPuXKs
9L5o1t7YvqRX25qJR3RLdpnRp6fTli7R1u6Qi/sqFjKv5nfbz+1MnB8LE8Wdt3Q1ac2ekLr/4Ydc
1tzimohla6VqA3YdmnvA9NGes/9Pr1+82nzKjIshU5L0VXI8bl6UXdp/vmNNHcfS1FXc+85N7mxj
3NH++IjWVguPZ8Vrv7zb5/EouuFHgk3z8eNCrtf+BVbP+NAic7h88c6K7UcTpp7Y/lasRkWrJDVI
YUv2026zsqjDKYdCAj1KXKq/2x1O2mssy5+sP5nr+Ap7UWm34/v3aTocYkvm3eZ8OfdIjNbz31cv
eNy78T3H7/4inVrZZ78nz9f4LWvx0H7p9mu3pwXFWudxdv6q5L8Z+65t+VnJA/O37uGwL+d+bjrl
rvk9ruVHDXSn91WdfjaPWWP1XwbBaw1WwtN85orfOt590HLKdluHr2E7Dj/onfdjVYnvaknr/Ohr
8afT53TY8iV38GRNvjfT1s0rbGqq2AOlRX1OLpLL9NaEmnSWtH7+7ZszzWNBeH1YUZKJTeXhfYl3
3t/lknoVeNmr9JHGjCKh9LUmhVsXcFYHLtnwKWTvBUed2WHpR1RZTZt+8Uxruhi7r6py4qnnt/te
Xc3N1HRerfDkQfXWtbGV5zNSPjzmD5nSeWTZ1u8/ZcWSNs8+AQA="}

I know I am feature creeping to hell right now so we may want to break this out as a follow up. I just want to find a way to avoid more breaking changes as we add more optional metadata to sign as it relates to a given set of code changes.

**Comment by lrvick** [comment]
Also: mac-comptibility was already merged. I think this needs to target master.

**Comment by KellerFuchs** [comment]
@lrvick I'm not sure it makes sense to stuff this in the same git note as the source review info.

**Comment by KellerFuchs** [comment]
Also, yes, please make a separate issue.

**Comment by lrvick** [comment]
@KellerFuchs If this metadata does not go in the same git note, then it is not all signed together. For projects that yeild a binary signing the binary with the commit hash review attests this code yeilds this hash. For non binary projects this would not be required. 

This sort of feature is totally out of scope, but I am making the case that we now have 2 different bits of data that need to be signed by each reviewer and more in the future, thus it probably makes sense to use a key/value structure leaving us room to expand in the future.

**Comment by tylerlevine** [comment]
> So I have been thinking a lot on this, and realized there is already yet one more thing we will want to potentially attest to in future revisons: The compiled binary hash this given patch-id produces.
> 
> So far the things that make sense to sign are:
> 
> * patch-id (always)
> * target head (optional)
> * compiled binary hash (optional)
> 
> There may yet be others we have not considered yet.
> 
> For this reason since we are making a breaking change to the signing format with this commit, let's tweak it to be future proof by using key/value pairs.
> 
> JSON sounds the easiest for this, but that will require we depend on jq (parsing json in bash in hell). Jq is becoming very common though so maybe this is acceptable until someone is willing to rewrite this whole thing as a Rust binary or somthing.
> 
> Anyway I was thinking a signature could then just be a single line of json containing a body and a signature for that body
> 
> {body:{ patchid:"somepatchid",bin-hash:"somebinhash",target-head:"targetheadhere"},signature:"owGbwMvMwMHY577wjenCUFnG00FJDNFzp3pWFySWJGdkplgpFefnpkI5SjpJmXm6GYnFGRBhIA/E
> UdIpSSxKTy3RzUhNBGqAcEDsjNSiVKVark4mYxYGRg4GWTFFlvRQ+71LsrdL7dXj5oZZysoEspGB
> i1MAJsIzkf8P37uGc1wx5SJH0kK6uGU+Lbu2WVtq9brA62/f5arGnBG4ct37pEa9++LuTZMPuXKs
> 9L5o1t7YvqRX25qJR3RLdpnRp6fTli7R1u6Qi/sqFjKv5nfbz+1MnB8LE8Wdt3Q1ac2ekLr/4Ydc
> 1tzimohla6VqA3YdmnvA9NGes/9Pr1+82nzKjIshU5L0VXI8bl6UXdp/vmNNHcfS1FXc+85N7mxj
> 3NH++IjWVguPZ8Vrv7zb5/EouuFHgk3z8eNCrtf+BVbP+NAic7h88c6K7UcTpp7Y/lasRkWrJDVI
> YUv2026zsqjDKYdCAj1KXKq/2x1O2mssy5+sP5nr+Ap7UWm34/v3aTocYkvm3eZ8OfdIjNbz31cv
> eNy78T3H7/4inVrZZ78nz9f4LWvx0H7p9mu3pwXFWudxdv6q5L8Z+65t+VnJA/O37uGwL+d+bjrl
> rvk9ruVHDXSn91WdfjaPWWP1XwbBaw1WwtN85orfOt590HLKdluHr2E7Dj/onfdjVYnvaknr/Ohr
> 8afT53TY8iV38GRNvjfT1s0rbGqq2AOlRX1OLpLL9NaEmnSWtH7+7ZszzWNBeH1YUZKJTeXhfYl3
> 3t/lknoVeNmr9JHGjCKh9LUmhVsXcFYHLtnwKWTvBUed2WHpR1RZTZt+8Uxruhi7r6py4qnnt/te
> Xc3N1HRerfDkQfXWtbGV5zNSPjzmD5nSeWTZ1u8/ZcWSNs8+AQA="}
> 
> I know I am feature creeping to hell right now so we may want to break this out as a follow up. I just want to find a way to avoid more breaking changes as we add more optional metadata to sign as it relates to a given set of code changes.

I'm not sure how useful that is unless your build system is deterministic. Can we break this out into a separate issue for further discussion?

**Comment by tylerlevine** [comment]
> ```
> $ ./bin/git-signatures init
> $ ./bin/git-signatures show
>  Public Key ID    | Status     | Trust     | Date                         | Signer Name                                        
> =======================================================================================================================
> error: no note found for object 0d0f7ac917b3bfae9b1f5ee308866494d9206d19.
> ```
> Re: signing key, indeed somehow only my rsa signing key got added...

You'll have to redo the rebase (the ref in `notes.rewriteRef` only gets moved when doing rebases or amends).

**Comment by lrvick** [comment]
> > So I have been thinking a lot on this, and realized there is already yet one more thing we will want to potentially attest to in future revisons: The compiled binary hash this given patch-id produces.
> > So far the things that make sense to sign are:
> > 
> > * patch-id (always)
> > * target head (optional)
> > * compiled binary hash (optional)
> > 
> > There may yet be others we have not considered yet.
> > For this reason since we are making a breaking change to the signing format with this commit, let's tweak it to be future proof by using key/value pairs.
> > JSON sounds the easiest for this, but that will require we depend on jq (parsing json in bash in hell). Jq is becoming very common though so maybe this is acceptable until someone is willing to rewrite this whole thing as a Rust binary or somthing.
> > Anyway I was thinking a signature could then just be a single line of json containing a body and a signature for that body
> > {body:{ patchid:"somepatchid",bin-hash:"somebinhash",target-head:"targetheadhere"},signature:"owGbwMvMwMHY577wjenCUFnG00FJDNFzp3pWFySWJGdkplgpFefnpkI5SjpJmXm6GYnFGRBhIA/E
> > UdIpSSxKTy3RzUhNBGqAcEDsjNSiVKVark4mYxYGRg4GWTFFlvRQ+71LsrdL7dXj5oZZysoEspGB
> > i1MAJsIzkf8P37uGc1wx5SJH0kK6uGU+Lbu2WVtq9brA62/f5arGnBG4ct37pEa9++LuTZMPuXKs
> > 9L5o1t7YvqRX25qJR3RLdpnRp6fTli7R1u6Qi/sqFjKv5nfbz+1MnB8LE8Wdt3Q1ac2ekLr/4Ydc
> > 1tzimohla6VqA3YdmnvA9NGes/9Pr1+82nzKjIshU5L0VXI8bl6UXdp/vmNNHcfS1FXc+85N7mxj
> > 3NH++IjWVguPZ8Vrv7zb5/EouuFHgk3z8eNCrtf+BVbP+NAic7h88c6K7UcTpp7Y/lasRkWrJDVI
> > YUv2026zsqjDKYdCAj1KXKq/2x1O2mssy5+sP5nr+Ap7UWm34/v3aTocYkvm3eZ8OfdIjNbz31cv
> > eNy78T3H7/4inVrZZ78nz9f4LWvx0H7p9mu3pwXFWudxdv6q5L8Z+65t+VnJA/O37uGwL+d+bjrl
> > rvk9ruVHDXSn91WdfjaPWWP1XwbBaw1WwtN85orfOt590HLKdluHr2E7Dj/onfdjVYnvaknr/Ohr
> > 8afT53TY8iV38GRNvjfT1s0rbGqq2AOlRX1OLpLL9NaEmnSWtH7+7ZszzWNBeH1YUZKJTeXhfYl3
> > 3t/lknoVeNmr9JHGjCKh9LUmhVsXcFYHLtnwKWTvBUed2WHpR1RZTZt+8Uxruhi7r6py4qnnt/te
> > Xc3N1HRerfDkQfXWtbGV5zNSPjzmD5nSeWTZ1u8/ZcWSNs8+AQA="}
> > I know I am feature creeping to hell right now so we may want to break this out as a follow up. I just want to find a way to avoid more breaking changes as we add more optional metadata to sign as it relates to a given set of code changes.
> 
> I'm not sure how useful that is unless your build system is deterministic. Can we break this out into a separate issue for further discussion?

Deterministic builds are becoming very much the standard for open source or security focused projects. You can't reliably assert that a given binary came from a given code from more than one location otherwise. But yeah we can let this merge and discuss a payload refactor to include generic metadata like this after.

I think this project is early enough that we can justify breaking changes easily enough.

**Comment by KellerFuchs** [comment]
@lrvick Please make a separate issue.

**Comment by lrvick** [comment]
CI keeps failing with BADSIG. Is there some some missing deps in the CI configs?

**Comment by tylerlevine** [comment]
It seems strange that it's fine on travis but failing on gitlab. Any idea what's wrong? I can validate the signatures locally.

**Comment by rafasc** [comment]
I think this is an acceptable alternative rafasc@f39eee0833f32f5d7b08100ebba02dfcf05fa089

**Comment by lrvick** [comment]
@rafasc You had dropped off IRC before I could thank you so thanks!

Sounds like minimum version of v2.11.0 should be safe here then, yeah?

**Comment by rafasc** [comment]
@lrvick feel free to cherry-pick that commit.

I guess with that the minimum version becomes v2.1.0.
Which is the version where `--stable` was introduced. You could go further back, but you would need to drop `--stable` and its properties.

**Comment by lrvick** [comment]
This PR was targeted against "mac-compatibility" instead of master. I merged to both branches to make this show as merged.
